ios/browser/api/favicon/favicon_driver.mm

```diff
@@ -13,60 +16,132 @@
   if ((self = [super init])) {
-    ios::ChromeBrowserStateManager* browser_state_manager =
-        GetApplicationContext()->GetChromeBrowserStateManager();
-    CHECK(browser_state_manager);
+    _webState = webState;
 
-    browser_state_ = browser_state_manager->GetLastUsedBrowserState();
-    CHECK(browser_state_);
+    web::WebState* real_web_state = [webState internalWebState].get();
+    DCHECK(real_web_state);
 
-    if (privateMode) {
-      browser_state_ = browser_state_->GetOffTheRecordChromeBrowserState();
-      CHECK(browser_state_);
-    }
+    ChromeBrowserState* original_browser_state =
+        ChromeBrowserState::FromBrowserState(real_web_state->GetBrowserState());
 
-    brave_favicon::BraveIOSWebFaviconDriver::CreateForBrowserState(
-        browser_state_,
+    brave_favicon::BraveIOSWebFaviconDriver::CreateForWebState(
+        real_web_state,
         ios::FaviconServiceFactory::GetForBrowserState(
-            browser_state_, ServiceAccessType::EXPLICIT_ACCESS));
+            original_browser_state, ServiceAccessType::EXPLICIT_ACCESS));
   }
   return self;
 }
 
+- (void)deinit {
+  web::WebState* real_web_state = [self.webState internalWebState].get();
+  DCHECK(real_web_state);
+
+  brave_favicon::BraveIOSWebFaviconDriver* driver =
+      brave_favicon::BraveIOSWebFaviconDriver::FromWebState(real_web_state);
+  DCHECK(driver);
+
+  if (driver_observer_) {
+    driver->RemoveObserver(driver_observer_.get());
+    driver_observer_ = nullptr;
+  }
+}
+
 - (void)setMaximumFaviconImageSize:(NSUInteger)maxImageSize {
+  web::WebState* real_web_state = [self.webState internalWebState].get();
+  DCHECK(real_web_state);
+
   brave_favicon::BraveIOSWebFaviconDriver* driver =
-      brave_favicon::BraveIOSWebFaviconDriver::FromBrowserState(browser_state_);
+      brave_favicon::BraveIOSWebFaviconDriver::FromWebState(real_web_state);
   DCHECK(driver);
   driver->SetMaximumFaviconImageSize(maxImageSize);
 }
 
 - (void)webView:(WKWebView*)webView
-    onFaviconURLsUpdated:(WKScriptMessage*)scriptMessage {
+       scriptMessage:(WKScriptMessage*)scriptMessage
+    onFaviconUpdated:(void (^)(NSURL*, UIImage*))callback {
+  web::WebState* real_web_state = [self.webState internalWebState].get();
+  DCHECK(real_web_state);
+
+  brave_favicon::BraveIOSWebFaviconDriver* driver =
+      brave_favicon::BraveIOSWebFaviconDriver::FromWebState(real_web_state);
+  DCHECK(driver);
+
+  if (driver_observer_) {
+    driver->RemoveObserver(driver_observer_.get());
+    driver_observer_ = nullptr;
+  }
+
   NSURL* ns_url = scriptMessage.frameInfo.request.URL;
   absl::optional<GURL> url;
   if (ns_url) {
     url = net::GURLWithNSURL(ns_url);
   }
 
-  web::ScriptMessage message(web::ValueResultFromWKResult(scriptMessage.body),
-                             false, scriptMessage.frameInfo.mainFrame, url);
+  web::ScriptMessage message =
+      web::ScriptMessage(web::ValueResultFromWKResult(scriptMessage.body),
```

Comment:
`scriptMessage.body` - is this produced by a custom JS we inject? If so, is that script sandboxed in `wkcontentworld`?

Reply:
Yes, that part is done on the iOS side, and not in the PR. So this PR doesn't care where it gets the message or body from, so long as it gets one.

The iOS side I sandboxed already in the iOS PR that will be updated after this merges.